servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java

```diff
@@ -59,17 +116,75 @@ public static synchronized void initializeFileSystem(boolean permissionCheckEnab
    * @param groups in which user belongs to
    * @param action requested {@link FileSystemAction} by user
    * @param inodes all the inodes retrieved by traversing the path
+   * @param selfIndex the index of the last inode in the inodes list
+   * @param parentOrAncestorIndex the index of the parent or ancestor inode in the inodes list
+   * @param isCheckOwner indicates whether to check that the user is the owner of the last inode
    * @throws IOException if permission checking fails
    */
-  public static void checkPermission(User user, List<String> groups, FileSystemAction action,
-      List<Inode> inodes) throws IOException {
+  private static void checkPermission(String user, List<String> groups, FileSystemAction action,
+      List<Inode> inodes, int selfIndex, int parentOrAncestorIndex, boolean isCheckOwner)
+      throws IOException {
+    int size = inodes.size();
+    assert size > 0;
+
     if (!sPermissionCheckEnabled) {
       return;
     }
 
-    if (sFileSystemOwner.equals(user.getName()) || groups.contains(sFileSystemSuperGroup)) {
+    if (sFileSystemOwner.equals(user) || groups.contains(sFileSystemSuperGroup)) {
```

Comment:
seeing as your current unit tests don't catch this problem, you should add a test for this